test/e2e/cert_manager_deployment_test.go

```diff
@@ -236,48 +236,62 @@ func TestCertRenew(t *testing.T) {
 			},
 		},
 	}
-	certmanagerv1Client, err := certmanagerclientset.NewForConfig(config)
-	require.NoError(t, err)
-	crt, err = certmanagerv1Client.CertmanagerV1().Certificates(crt.ObjectMeta.Namespace).Create(ctx, crt, metav1.CreateOptions{})
-	defer certmanagerv1Client.CertmanagerV1().Certificates(crt.ObjectMeta.Namespace).Delete(ctx, crt.ObjectMeta.Name, metav1.DeleteOptions{})
-	require.NoError(t, err)
-	err = wait.PollImmediate(PollInterval, TestTimeout, func() (bool, error) {
+	// certmanagerClient is defined in utils_test and reusing it here for waitForCertificateReadiness().
+	certmanagerClient, err = certmanagerclientset.NewForConfig(config)
+	require.NoErrorf(t, err, "failed to create cert manager client: %v", err)
+	_, err = certmanagerClient.CertmanagerV1().Certificates(crt.Namespace).Create(ctx, crt, metav1.CreateOptions{})
+	require.NoError(t, err, "failed to create certificate: %v", err)
+	defer certmanagerClient.CertmanagerV1().Certificates(crt.Namespace).Delete(ctx, crt.Name, metav1.DeleteOptions{})
+
+	err = waitForCertificateReadiness(ctx, crt.Name, crt.Namespace)
+	require.NoError(t, err, "failed while waiting for certificate to be ready: %v", err)
+
+	crt, err = certmanagerClient.CertmanagerV1().Certificates(crt.Namespace).Get(ctx, crt.Name, metav1.GetOptions{})
+	require.NoError(t, err, "failed to fetch certificate: %v", err)
+
+	// Store the NotAfter validity time of the newly created certificate.
+	expiryTime := crt.Status.NotAfter.Time.UTC()
+	t.Logf("newly created Certificate's NotAfter: %v", expiryTime)
+
+	err = wait.PollUntilContextTimeout(ctx, PollInterval, TestTimeout, true, func(ctx context.Context) (bool, error) {
 		secret, _ := loader.KubeClient.CoreV1().Secrets(ns.Name).Get(ctx, crt.Spec.SecretName, metav1.GetOptions{})
 		tlsConfig, isValid := library.GetTLSConfig(secret)
 		if !isValid {
 			return false, nil
 		}
 
-		is_host_correct, err := library.VerifyHostname(ingress_host, tlsConfig.Clone())
+		isHostCorrect, err := library.VerifyHostname(ingressHost, tlsConfig.Clone())
 		if err != nil {
 			t.Errorf("Host %v", err)
 			return false, nil
 		}
-		is_not_expired, err := library.VerifyExpiry(ingress_host+":443", tlsConfig.Clone())
+
+		isNotExpired, err := library.VerifyExpiry(ingressHost+":443", tlsConfig.Clone())
 		if err != nil {
 			t.Errorf("Expiry %v", err)
 			return false, nil
 		}
-		expiryTime, err := library.GetCertExpiry(ingress_host+":443", tlsConfig.Clone())
-		t.Logf("Expiry Before %v", expiryTime)
-		if err != nil {
-			return false, nil
-		}
+
 		time.Sleep(time.Minute + time.Second*5)
```

Comment:
Optimization is required to wait for exactly the remaining time, using time.After() might help

Reply:
Let me check this, I don't think it's required, since polling is used.

Reply:
`time.Sleep(time.Minute + time.Second*5)` needs refinement.

Reply:
suggestion incorporated.